### ISSUE TITLE
Distinguish between NotFound resource and NotFound route

### ISSUE THREAD

**Issue by NullVoxPopuli** [json-api-dotnet/JsonApiDotNetCore#341]
does it make sense to return JSONAPI errors on uncaught server errors?

for example, I'm getting a 404, cause I'm bad at databasing, and the response body is empty.
imo, it'd be great if there was a response body with the error message text / and or exception class.

maybe this'd be a development mode only thing? idk.

**Comment by jaredcnance** [comment]
All errors should be returned in a json:api error document. Otherwise how would a json:api client know how to handle them? 

Can you provide more details around your specific situation? You’re getting a 404 because you’re returning null from your controller? Or an exception isn’t being handled properly?

**Comment by NullVoxPopuli** [comment]
I'm really just wanting _anything_ returned. lol.

Here is what I have so far:

![image](https://user-images.githubusercontent.com/199018/42701763-719d6b34-8695-11e8-9fd5-54afbaa873f9.png)

```
api_1      | [16:04:58 INF] Request starting HTTP/1.1 GET http://api:7081/api/users/current-user  
api_1      | [16:04:58 INF] Route matched with {action = "GetAsync", controller = "Users"}. Executing action Optimajet.DWKit.StarterApplication.Controllers.UsersController.GetAsync (Optimajet.DWKit.StarterApplication)
api_1      | [16:04:58 INF] Entity Framework Core 2.1.0-rtm-30799 initialized 'AppDbContext' using provider 'Npgsql.EntityFrameworkCore.PostgreSQL' with options: None
api_1      | [16:04:58 INF] Executing action method Optimajet.DWKit.StarterApplication.Controllers.UsersController.GetAsync (Optimajet.DWKit.StarterApplication) with arguments (["0"]) - Validation state: Invalid
api_1      | [16:04:58 INF] Executed DbCommand (1ms) [Parameters=[@__id_0='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
api_1      | SELECT e."Id", e."Email", e."ExternalId", e."Name"
api_1      | FROM "User" AS e
api_1      | WHERE e."Id" = @__id_0
api_1      | LIMIT 2
api_1      | [16:04:58 INF] Executed action method Optimajet.DWKit.StarterApplication.Controllers.UsersController.GetAsync (Optimajet.DWKit.StarterApplication), returned result Microsoft.AspNetCore.Mvc.NotFoundResult in 12.681ms.
api_1      | [16:04:58 INF] Executing HttpStatusCodeResult, setting HTTP status code 404
api_1      | [16:04:58 INF] Executed action Optimajet.DWKit.StarterApplication.Controllers.UsersController.GetAsync (Optimajet.DWKit.StarterApplication) in 18.5098ms
api_1      | [16:04:58 INF] Request finished in 27.8235ms 404 
```

**Comment by jaredcnance** [comment]
Does a user with the id “current-user” exist in the database? I don’t see a custom repository that would alter the default behavior. And by default it will just look for a user with the specified id in the database.

**Comment by NullVoxPopuli** [comment]
there is no "current-user" user atm, 
but, I'm more just wondering if there was any way to get more information in the response payload?

**Comment by jaredcnance** [comment]
Per the specification:
> A server MUST respond with 404 Not Found when processing a request to fetch a single resource that does not exist

However, you can always alter this behavior by overriding the controller method or implementing the necessary service or repository interfaces:

https://json-api-dotnet.github.io/#/custom-errors

https://json-api-dotnet.github.io/#/layer-overview

**Comment by jaredcnance** [comment]
You can always write your own middleware. But, I’m general, the most detailed information about an error is available at the site where the error occurs. It would be difficult to provide useful information using generalized middleware. However, for this example, you could create a general controller that overrides the GetAsync methods and returns an error body instead of just a 404.

**Comment by jaredcnance** [comment]
@NullVoxPopuli I'm closing this because I'm not sure I have any work to takeaway from this. Feel free to reopen if you're interested in continuing the discussion.

**Comment by jaredcnance** [comment]
re-opening because I want to track a solution for distinguishing between 404 route not found and 404 resource (with id) not found.

**Comment by bart-degreed** [comment]
Fixed by #714.
